Vector

td2r opened the pull request.

Commit:
first commit to vector

Commit:
vector space alignment

Commit:
vector another small space alignment

kt-cpp reviewed (changes requested):
Бан по причине скат (логические/семантические ошибки при этом присуствуют, но пока что опущены)

kt-cpp commented on vector/vector.h:
```diff
@@ -41,22 +49,327 @@ struct vector
+    template <typename T>
+    typename std::enable_if<std::is_trivially_destructible<T>::value>::type
+    destroy_all(T* dest, size_t size) {}
+
+    template <typename T>
+    typename std::enable_if<!std::is_trivially_destructible<T>::value>::type
+    destroy_all(T* dest, size_t size)
+    {
+        for (size_t i = 0; i < size; ++i) {
+            dest[i].~T();
+        }
+    }
+
+    template <typename T>
+    typename std::enable_if<std::is_nothrow_constructible<T>::value>::type
+    copy_construct_all(T* dest, T const* src, size_t size)
+    {
+        size_t i = 0;
+        while (i != size) {
+            new (dest + i) T(src[i]);
+            ++i;
+        }
+    }
+
+    template <typename T>
+    typename std::enable_if<!std::is_nothrow_constructible<T>::value>::type
+    copy_construct_all(T* dest, T const* src, size_t size)
+    {
+        size_t i = 0;
+        try {
+            while (i != size) {
+                new (dest + i) T(src[i]);
+                ++i;
+            }
+        } catch (...) {
+            destroy_all(dest, i);
+            throw;
+        }
+    }
+}
```
1. Лучше вызывать деструкторы в обратном порядке (т.е. как обычно)
2. Почитай, пойми что неправ, исправь
https://en.cppreference.com/w/cpp/types/is_copy_constructible
https://en.cppreference.com/w/cpp/types/is_constructible
3. Anonymous namespace in header

kt-cpp replied:
(про 3: хоть у нас и реализация функций в хедере, просто на будущее - не думай что можешь так писать)

td2r replied:
> (про 3: хоть у нас и реализация функций в хедере, просто на будущее - не думай что можешь так писать)

Это зарезолвленный комментарий или нет?
Просто написать sfinae на методах у меня не получилось. В шаблонных классах вовсе нельзя использовать anonymous namespace или это просто напоминание, что его нужно писать в .cpp?

kt-cpp replied:
Его нет смысла писать, если эти шабонные классы написаны в хедере. Потому что анонимный неймспейс в хедере - это тоже самое не писать его вообще, ибо всё из него будет видно

td2r replied:
>     1. Лучше вызывать деструкторы в обратном порядке (т.е. как обычно)
> 
>     2. Почитай, пойми что неправ, исправь
>        https://en.cppreference.com/w/cpp/types/is_copy_constructible
>        https://en.cppreference.com/w/cpp/types/is_constructible
> 
>     3. Anonymous namespace in header

1. Исправил
2. Исправил
3. Исправил - сделал static методами.

kt-cpp commented on vector/vector.h:
```diff
@@ -41,22 +49,327 @@ struct vector
+	return r;
+}
+
+template <typename T>
+typename vector<T>::iterator vector<T>::erase(const_iterator pos)
+{
+	auto p = const_cast<iterator>(pos);
+	iterator it = p;
+	++it;
+	iterator e = end();
+	while (it != e) {
+		std::swap(*(it - 1), *it);
+		++it;
+	}
+	pop_back();
+	return p;
+}
```
```erase(pos, pos +1)```

td2r replied:
> `erase(pos, pos +1)`

Fixed.

kt-cpp commented on vector/vector.h:
```diff
@@ -41,22 +49,327 @@ struct vector
+
+template <typename T>
+typename vector<T>::iterator vector<T>::erase(const_iterator first, const_iterator last)
+{
```
По поводу ассертов
1. Ты либо пишешь их везде (в таких функциях)
2. Либо не пишешь, и считаешь это за UB
Т.е. например front() на std::vector

Ассерты созданы для того, чтобы ты мог проверять свои собственные инварианты

td2r replied:
> По поводу ассертов
> 
>     1. Ты либо пишешь их везде (в таких функциях)
> 
>     2. Либо не пишешь, и считаешь это за UB
>        Т.е. например front() на std::vector
> 
> 
> Ассерты созданы для того, чтобы ты мог проверять свои собственные инварианты

Fixed: добавил ассерты в insert и erase(first, last).

kt-cpp commented on vector/vector.h:
```diff
@@ -41,22 +49,327 @@ struct vector
+	iterator r = tmp.end();
+	tmp.push_back(element);
+	copy_construct_all(tmp.end(), p, end() - p);
+	tmp.size_ += end() - p;
```
Про ассерты написано в другом сообщении, но это хороший пример

Например, если бы таких строчек было 100, то в конце хорошо бы сделать ```assert(tmp.size_ == size_ + 1)```

td2r replied:
> Про ассерты написано в другом сообщении, но это хороший пример
> 
> Например, если бы таких строчек было 100, то в конце хорошо бы сделать `assert(tmp.size_ == size_ + 1)`

Добавил.

kt-cpp commented on vector/vector.h:
```diff
@@ -41,22 +49,327 @@ struct vector
+
+template <typename T>
+void vector<T>::swap(vector& rhs)
+{
```
Вот здесь можно добавить ```using std::swap;```

И вызвать ```swap``` вместо ```std::swap```

Тогда получишь доступ к другим возможным перегрузкам

td2r replied:
> Вот здесь можно добавить `using std::swap;`
> 
> И вызвать `swap` вместо `std::swap`
> 
> Тогда получишь доступ к другим возможным перегрузкам

Fixed.

kt-cpp reviewed (changes requested):
Fix please

Commit:
vector fix1
1. destroy_all now calls destructors in backward order
2. is_nothrow_constructible replaced with is_nothow_copy_constructible in copy_construct_all
3. erase(pos) simplified with erase(pos, pos + 1)
4. added asserts in insert and erase(first, last) (erase(pos) use erase(first, last))
5. using std::swap in vector::swap

Commit:
vector fix1
6. destroy_all and copy_construct_all changed to static private methods instead of anonymous namespace functions

kt-cpp commented on vector/vector.h:
```diff
@@ -41,22 +49,360 @@ struct vector
+    if (other.size_ != 0) {
+        data_ = static_cast<T*>(operator new(other.size_ * sizeof(T)));
+        try {
+            copy_construct_all(data_, other.data_, other.size_);
```
[note] Ты сделал сфинае перегрузки, но при этом тут всё равно ловишь исключение. Можно конечно поверить в компилятор, который выкинет `try`, но я верю, что заморочившись можно лучше

td2r replied:
Действительно, не обратил внимания.
Сделал при помощи unique_ptr:
если copy_construct_all кидает exception, то при выходе из области видимости удалится массив
иначе просто освобождаю указатель и присваиваю в data_.

kt-cpp replied:
Интересное решение

kt-cpp commented on vector/vector.h:
```diff
@@ -41,22 +49,360 @@ struct vector
+void vector<T>::push_back(T const& element)
+{
+    if (size_ == capacity_) {
+        vector<T> tmp;
+        tmp.capacity_ = increase_capacity();
+        tmp.data_ = static_cast<T*>(operator new(tmp.capacity_ * sizeof(T)));
```
[minor] Почему бы не вызвать просто у него `reserve`?

td2r replied:
Fixed.

kt-cpp commented on vector/vector.h:
```diff
@@ -41,22 +49,360 @@ struct vector
+}
+
+template <typename T>
+typename vector<T>::iterator vector<T>::insert(const_iterator pos, T const& element)
+{
+    assert(begin() <= pos && pos <= end());
+
+    auto p = const_cast<iterator>(pos);
+    vector<T> tmp;
+    tmp.new_buffer(size_ == capacity_ ? increase_capacity() : capacity_);
+
+    copy_construct_all(tmp.data_, data_, p - begin());
+    tmp.size_ = p - begin();
+
+    iterator r = tmp.end();
+    tmp.push_back(element);
+
+    copy_construct_all(tmp.end(), p, end() - p);
+    tmp.size_ += end() - p;
+
+    assert(tmp.size_ == size_ + 1);
+
+    swap(tmp);
+    return r;
+}
```
В целом не нравится создание временного вектора и копирование абсолютно всех элементов. Хочется в отсутствии реаллокаций работать за расстояние от конца до `pos`. Подумай как это, если что можем подсказать

td2r replied:
Похоже я зачем-то решил повысить гарантии. Но поскольку обещаем weak, то можно переписать на end() - pos свопов, что и сделал.

kt-cpp replied:
Если верить, что `swap` noexcept (что почти всегда правда), у элементов, то мы вполне себе получаем тут строгую гарантию, ибо её имеет `push_back`

kt-cpp commented on vector/vector.h:
```diff
@@ -41,22 +49,360 @@ struct vector
+typename vector<T>::iterator vector<T>::erase(const_iterator first, const_iterator last)
+{
+    assert(begin() <= first && last <= end());
+    auto p = const_cast<iterator>(first);
+    iterator p1 = p;
```
1. Зачем вообще существует `p`?)
2. Понимаешь ли ты почему эти `const_cast` тут валидны?

td2r replied:
1. Этот метод возвращает итератор на элемент, следующий за последним удаленным элементом. После всех свопов информация об этом итераторе в p1, p2 теряется, поэтому я его запоминаю и возвращаю.
2. Да, const_cast валиден, поскольку все элементы в нашем векторе не имеют идентификации const.
То есть такой код:
int const x = 2;
auto p = const_cast<int*>(&x);
*p = 3;
это undefined behaviour, поскольку x изначально создан как const и мог быть помещен в read-only секцию. Наши же объекты созданы как изменяемые.

kt-cpp commented on vector/vector.h:
```diff
@@ -41,22 +49,360 @@ struct vector
+}
+
+template <typename T>
+size_t vector<T>::increase_capacity() const
```
[minor] Кажется имя это функция не соответствует содержанию. Она не увеличивает вместимость, а лишь возвращает увеличенную

td2r replied:
Поменял на increased_capacity.

kt-cpp commented on vector/vector.h:
```diff
@@ -41,22 +49,360 @@ struct vector
+typename std::enable_if<std::is_nothrow_copy_constructible<T_>::value>::type
+vector<T>::copy_construct_all(T *dest, const T *src, size_t size)
+{
+    size_t i = 0;
+    while (i != size) {
+        new (dest + i) T(src[i]);
+        ++i;
+    }
```
[note] А чем `for` не угодил?

td2r replied:
Fixed.

kt-cpp commented on vector/vector.h:
```diff
@@ -41,22 +49,360 @@ struct vector
+{
+    size_t i = 0;
+    try {
+        while (i != size) {
+            new(dest + i) T(src[i]);
+            ++i;
+        }
```
[note] Тут тоже вопрос про `for`

td2r replied:
Fixed.

kt-cpp commented on vector/vector.h:
```diff
@@ -41,22 +49,360 @@ struct vector
+        destroy_all(dest, i);
+        throw;
+    }
+}
```
[note] Даже github подсказывает, что забыт перенос строки в конце файла

td2r replied:
Fixed.

Commit:
vector fix2
1. copy constructor now don't catch exception if T is nothrow copy constructible
2. simplify in push_back: tmp.reserve
3. insert rework: now it works in time end() - pos and swaps elements
4. small fixes in erase(first, last)
5. increase_capacity renamed to increased_capacity
6. copy_construct_all in is_nothrow_copy_constructible spec use for instead of while
7. \n in eof

Commit:
vector fix2
8. for instead of while in copy_construct_all in !is_nothrow_copy_constructible spec
